_delphi_utils_python/delphi_utils/logger.py

```diff
@@ -84,10 +119,116 @@ def get_structured_logger(name=__name__,
     system_logger = logging.getLogger(name)
     if filename and not system_logger.handlers:
         system_logger.addHandler(logging.FileHandler(filename))
-    system_logger.setLevel(logging.INFO)
+    system_logger.setLevel(log_level)
     logger = structlog.wrap_logger(system_logger)
 
     if log_exceptions:
         handle_exceptions(logger)
 
     return logger
+
+
+
+
+class LoggerThread():
+    """
+    the bare structlog loggers are thread-safe but not multiprocessing-safe.
+    a `LoggerThread` will spawn a thread that listens to a mp.Queue
+    and logs messages from it with the provided logger,
+    so other processes can send logging messages to it
+    via the logger-like `SubLogger` interface.
+    the SubLogger even logs the pid of the caller.
+
+    this is good to use with a set of jobs that are part of a mp.Pool,
+    but isnt recommended for general use
+    because of overhead from threading and multiprocessing,
+    and because it might introduce lag to log messages.
+    """
+
+    class SubLogger():
+        """
+        Multiprocessing-safe logger-like interface
+        to convey log messages to a listening LoggerThread.
+        """
+
+        def __init__(self, queue):
+            self.queue = queue
+
+        def _log(self, level, *args, **kwargs):
+            kwargs_plus = {'sub_pid': multiprocessing.current_process().pid}
+            kwargs_plus.update(kwargs)
+            self.queue.put([level, args, kwargs_plus])
+
+        def info(self, *args, **kwargs):
+            """Log an INFO level message."""
+            self._log(logging.INFO, *args, **kwargs)
+
+        def warn(self, *args, **kwargs):
+            """Log a WARN level message."""
+            self._log(logging.WARN, *args, **kwargs)
+
+
+    def get_sublogger(self):
+        """Accessor method to retrieve a SubLogger for this LoggerThread."""
+        return self.sublogger
+
+    def __init__(self, logger, q=None):
+        self.logger = logger
+        if q:
+            self.msg_queue = q
+        else:
+            self.msg_queue = multiprocessing.Queue()
+
+        def logger_thread_worker():
+            logger.info('thread started')
+            while True:
+                msg = self.msg_queue.get()
+                if msg == 'STOP':
+                    logger.debug('received stop signal')
+                    break
+                level, args, kwargs = msg
+                if level == logging.INFO:
+                    logger.info(*args, **kwargs)
+                if level == logging.WARN:
+                    logger.warn(*args, **kwargs)
```

Comment:
Do we need debug or exception (or others) here too? Why/why not?

Reply:
we only need the methods we use, and right now, nobody is using debug or exception.  i had a TODO here to make it more obvious to anyone who came looking that they could add their own, but the current pylint settings disallowed that...

this is sort of a proxy or facade for a Logger, and it doesnt need to duplicate the whole interface.  someone using a SubLogger got there by providing an existing Logger object in the first place, so hopefully they have some idea that its still wrapped up in there.

Reply:
...i caved and added some other levels